### DISCUSSION TITLE
SMTP Failed

### DISCUSSION THREAD

**Discussion by afanjul** [postalserver/postal#1438] [Help with using Postal]
Hi there, we are having a strange behaviour with the SMTP service of the Postal server. The SMTP service is shown as Failed after 5 retries, but the server is working "fine" anyway (at least it sends emails). We don't have any other service in that port (we uninstalled postfix in the very beginning) and netstat does not show anything strange.

Postal status:
![image](https://user-images.githubusercontent.com/2046424/89036485-12cb4a80-d33d-11ea-8ea1-4ffb1908fc93.png)

Netstat:
![image](https://user-images.githubusercontent.com/2046424/89036629-51f99b80-d33d-11ea-9bdc-275e0585c7e0.png)

Postal smtp_server.log:
![image](https://user-images.githubusercontent.com/2046424/89036671-6342a800-d33d-11ea-9506-357052f381c8.png)

Postal smtp_sender.log:
![image](https://user-images.githubusercontent.com/2046424/89036749-853c2a80-d33d-11ea-90b6-159ce87e5190.png)

-----------------
Any ideas about why we are having a Failed SMTP service???
Thanks

**Comment by willpower232** [comment]
The SMTP server part is for incoming email

Do you have an `smtp_server.log` file? that might give you a clue.

Failing that, you can also try running `bundle exec rake postal:smtp_server` as the postal user to see the error happen in real time.

**Comment by afanjul** [comment]
Sorry @willpower232 I didn't see your message before: 
It's pretty strange that the smtp server is been used for ¿incoming email? the SMTP servers are usually used for sending emails, otherwise I already attached the smtp_server.log screenshot in my first comment (with the typical command for outcoming emails). 
The `bundle exec rake postal:smtp_server` command throws an error: `bundler: failed to load command: rake (/usr/bin/rake)
Gem::GemNotFoundException: can't find gem rake (>= 0.a)
  /usr/lib/ruby/2.3.0/rubygems.rb:241:in `bin_path'
  /usr/bin/rake:22:in `<top (required)>'`

It's a production environment so i'm bit afraid of installing new gems or something.

**Comment by willpower232** [comment]
are you running that command as the postal user in the postal directory?

**Comment by afanjul** [comment]
@willpower232 due to your last message I tried to run "postal start" as postal user (using `sudo su - postal` before), and now the smtp keeps running without "failed". So it seems a conflict with user permissions or something... 

![image](https://user-images.githubusercontent.com/2046424/94912957-b0abd480-04a8-11eb-93bc-54acaabe7e39.png)


What is the best practices to launch postal? as root user or postal user?

thanks

**Comment by willpower232** [comment]
IIRC `sudo su` doesn't always work exactly as you think it should, I would try running `sudo -iu postal` to switch to the postal user fully and then `bundle exec rake postal:smtp_server`

**Comment by uhlhosting** [comment]
```
bundle exec rake postal:smtp_server --trace
** Invoke postal:smtp_server (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute postal:smtp_server
rake aborted!

Errno::EADDRINUSE: Address already in use - bind(2) for "::" port 25

/opt/postal/app/lib/postal/smtp_server/server.rb:52:in `initialize'
/opt/postal/app/lib/postal/smtp_server/server.rb:52:in `open'
/opt/postal/app/lib/postal/smtp_server/server.rb:52:in `listen'
/opt/postal/app/lib/postal/smtp_server/server.rb:258:in `run'
/opt/postal/app/lib/tasks/postal.rake:13:in `block (2 levels) in <top (required)>'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/task.rb:273:in `block in execute'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/task.rb:273:in `each'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/task.rb:273:in `execute'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/task.rb:214:in `block in invoke_with_call_chain'
/usr/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/task.rb:194:in `invoke_with_call_chain'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/task.rb:183:in `invoke'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/application.rb:160:in `invoke_task'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/application.rb:116:in `each'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/application.rb:116:in `block in top_level'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/application.rb:125:in `run_with_threads'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/application.rb:110:in `top_level'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/application.rb:83:in `block in run'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/application.rb:186:in `standard_exception_handling'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/lib/rake/application.rb:80:in `run'
/opt/postal/vendor/bundle/ruby/2.3.0/gems/rake-12.3.3/exe/rake:27:in `<top (required)>'
/opt/postal/vendor/bundle/ruby/2.3.0/bin/rake:22:in `load'
/opt/postal/vendor/bundle/ruby/2.3.0/bin/rake:22:in `<main>'

Tasks: TOP => postal:smtp_server
```


Got this when I try to debug SMTP.

**Comment by willpower232** [comment]
> Errno::EADDRINUSE: Address already in use - bind(2) for "::" port 25

Either you're already running Postal or something else is using that port. You can verify with `netstat -lnp`

**Comment by dragoangel** [comment]
`Errno::EAFNOSUPPORT: Address family not supported by protocol - socket(2) for "::" port 25`
This error is interesting. On ubuntu with disabled IPv6. Fixed temporary by changing `/opt/postal/app/lib/postal/smtp_server/server.rb`:
`@server = TCPServer.open('::', Postal.config.smtp_server.port)` to `@server = TCPServer.open('0.0.0.0', Postal.config.smtp_server.port)`
But this really strange...

**Comment by willpower232** [comment]
Not super strange to be honest, `'::'` is the term for all IPv6 and IPv4 addresses, if you have disabled IPv6 then you have to use the IPv4 version `'0.0.0.0'`

**Comment by dragoangel** [comment]
@willpower232 I agree, but there need to be fallback retry to ipv4 (or check what proto is supported) by default if `::` isn't working ;), no?

**Comment by willpower232** [comment]
Removing the servers ability to use IPv6 is a drastic customisation that is not common and should not be required in an overwhelming majority of cases. See also #813

**Comment by dragoangel** [comment]
Not agree, many companies has security hardering that disables ipv6 where it not must be in place (not exists in lan). So fallback is common for local binding and I think it not hard to write Rescue Exceptions here :)

P.s.: I also curious:
1. why postal not have binding parameter in config to which ip need to bind smtp server?
2. why postal not support multiple ports?;) This was asked many many people here's. Now only one way to get this is do NAT by yourself on localhost. To note: security hardering of some companies also restrictions usage of NAT on non gw machines :).
3. postal has text in web ui that tells to use 25 or 2525 port to send mail. This incorrect as postal actually open connection on 25 port only ;). if you change this behavior to allow multiple ports bindings you can also display correct help info in web ui, e.g.: `html template config.smtp_ser.ports`
4. Why postal suggests to people use uncommon 2525 port as client connection point? This done by standard on submission 587 port. And 25 port at all mustn't be used for client-to-MTA (postal) connection as this port used only for MTA-to-MTA communication and in general all outgoing connections for non-MTA server to 25 port must be denied to prevent any possible spam 😊.

E.g. pseudocode logic of utilizing multiply IP:ports connections from config can looks like this:
```
config.smtp_server.ips=0.0.0.0,:: #default
config.smtp_server.ports=25,587 #default
server=0;
failed_server=0;
foreach ip in ips do:
    foreach port in ports do:
        begin
             server++;
             TCP server open...
        rescue
             failed_server++;
             log(warn)"cant bind ip:port";
        finally
             if server == failded_server do:
                 log(fatal)"no ip:ports can be open..."
                 exit 1 # here all postal actually can be shut-downed as mail server actually not working at all :D
         end
   end
end
```
